### PR TITLE
Change STF to work on block level instead of batch level

### LINKED ISSUE #408: Based Sequencing with Soft Confirmations
## Background
The module system currently has work in-progress to enable "based sequencing". That will be great for censorship resistance, but it sacrifices the sequencer's ability to provide near-instant soft confirmations. Since no sequencer has a "lock" on global state, they can't guarantee that no other transaction will come in and mutate the state that they depended on. 

For example, suppose that Alice wants to make a trade on Uniswap swapping one $ETH for 1000 $DAI with a slippage tolerance of 1%. Accordingly, she signs a transaction and sends it to a sequencer, Steve, who gives her a soft-confirmation that the trade will go through. At about the same time, Bob sends a transaction to another sequencer Tom, which swaps 1000 $ETH for $DAI, moving the price by 30%. Now, if Tom's block lands on the DA layer in front of Steve's the soft confirmation for Alice will be voided. 

To provide a competitive UX, we would very much like to fix that flaw. 

## Naive Implementation: Based Sequencing with "Preferred Sequencer"
A naive approach to this problem would be to enshrine some centralized sequencer and guarantee that they will always get first execution rights in each DA layer slot. So, instead of processing blobs in _exactly_ the order they appear on the DA layer, the rollup always processes the blob from the preferred sequencer as if it had appeared at the first index in its slot. 

Unfortunately, this approach is vulnerable to censorship. To be concrete, suppose that at the end of DA slot 100 there's a significant arbitrage between price of $ETH on two Uniswap `v2` vs `v3`. By default, the preferred sequencer will get to extract this arbitrage, since their blob will be processed as if it had appeared at the very beginning of slot 101. But a smart DA layer block proposer will want to extract the arbitrage for themself. So, they'll have an incentive to create their own blob which extracts the arbitrage, and censor the preferred sequencer to prevent their blob from being processed. 

## Getting Smarter: Simulating Multiple Proposers

To eliminate the incentive for censorship, we need to ensure that the DA layer block proposer can't get their transaction processed if they censor the Preferred Sequencer. One possible approach would be to return to centralized sequencing, but that is undesirable because of the risk of censorship. A smarter approach is to require multiple DA layer block proposers to collude in order to censor effectively, while simultaneously encouraging each individual proposer to defect from the cartel. In other words, we want to ensure that non-censorship is the dominant strategy in block proposal creation. 

To do that, we need to alter the rollup's state transition rule. Previously, our proposed rule looked like this:
```rust
let blobs_to_process = vec![];
// first, apply the blob from the preferred sequencer
for blob in blobs {
   if blob.sender() == PREFERRED_SEQUENCER {
     rollup.apply_blob(blob);
  } else { 
     blobs_to_process.push(blob)
  } 
}
//  then, apply all the other blobs
for blob in blobs_to_process {
   rollup.apply_blob(blob);
}
```

We can alter that rule to defer execution of blobs from non-preferred sequencers for an even longer period. Instead of processing blobs from other sequencers within the current slot, we can save them (or, in practice, a commitment to them) in our proof output, and process them as part of the *next slot

```rust
let blobs_to_defer = vec![];
// first, apply the blob from the preferred sequencer
for blob in blobs {
   if blob.sender() == PREFERRED_SEQUENCER {
     rollup.apply_blob(blob);
  } else { 
     blobs_to_process.push(blob)
  } 
}

let deferred_blob_root = make_merkle_tree(blobs_to_defer);
let blobs_to_process = Zkvm::read_from_host();
ensure!(make_merkle_tree(blobs_to_process) == old_deferred_blob_root);

for blob in blobs_to_process {
   rollup.apply_blob(blob);
}
Zkvm::commit(deferred_blob_root);
```

This way, even if the current block proposer censors the preferred sequencer and inserts his own arbitrage transaction onto the DA layer, his transaction won't be processed until the end of the next slot. And since the next DA layer proposer doesn't have anything to gain by censoring the preferred sequencer, we can expect that the sequencer's bundle will land in that slot and extract the arbitrage. 

We can carry this idea even further by deferring blob execution for an even longer time. Instead of processing "based" blobs in the next slot, we can process them 3 or 5 blocks in the future. The more sequencers will have to collude in order to extract their own arbitrage - and the less incentive any of them will have to do so.

### PR DESCRIPTION
**Pull request by citizen-stig**
# Summary of Proposed Changes

We propose a significant refactor to improve the logic and efficiency of how blobs are processed in each block, specifically in the context of sequencing with soft confirmation. These changes impact several key components and introduce new types. This write-up will detail the adjustments and their motivation, aiming for clarity and comprehension.

## Changes to `StateTransitionFunction`

I've condensed the three existing methods - `begin_slot`, `apply_blob`, and `end_slot` - into a single method called `apply_slot`. This change hands over the block handling logic to individual StateTransitionFunction (STF) implementations.

New design for `StateTransitionFunction` is as follows:

```rust
pub trait StateTransitionFunction<Vm: Zkvm, B: BlobReaderTrait> {
    type StateRoot;
    type TxReceiptContents: Serialize + DeserializeOwned + Clone;
    type BatchReceiptContents: Serialize + DeserializeOwned + Clone;
    type Witness: Default + Serialize + 'static;
    
    fn apply_slot<'a, I>(
        &mut self,
        witness: Self::Witness,
        blobs: I,
    ) -> SlotResult<
        Self::StateRoot,
        Self::BatchReceiptContents,
        Self::TxReceiptContents,
        Self::Witness,
    >
    where
        I: IntoIterator<Item = &'a mut B>;
}
```

###  Removing `MisbehaviorHint`

As it is currently unused and makes changes of interface more convoluted, we've decided to remove it for now,  and introduce it as part of the `Witness` type

### Introduction of `SlotResult`

`SlotResult` is a new container type, designed to streamline the output from the previous methods `apply_blob` and `end_slot`.

Previously, these methods looked like this:

```rust

    fn apply_blob(
        &mut self,
        blob: &mut B,
        misbehavior_hint: Option<Self::MisbehaviorProof>,
    ) -> BatchReceipt<Self::BatchReceiptContents, Self::TxReceiptContents>;

    fn end_slot(&mut self) -> (Self::StateRoot, Self::Witness);
```

Now, I introduce `SlotResult`:

```rust
///  - S - generic for state root
///  - B - generic for batch receipt contents
///  - T - generic for transaction receipt contents
///  - W - generic for witness
pub struct SlotResult<S, B, T, W> {
    pub state_root: S,
    pub batch_receipts: Vec<BatchReceipt<B, T>>,
    pub witness: W,
}
```

I decided to introduce `SlotResult` instead of using a type alias for a three-element tuple to increase readability.

### `BlobTransactionTrait` renamed to `BlobReaderTrait`

I've renamed `BlobTransactionTrait` to `BlobReaderTrait` to better reflect its actual role. This trait doesn't carry any transaction-specific aspects and might be used for Attestations or Proofs in the future.

### 'static bound for some traits

Traits like `BlobReaderTrait`, `AddressTrait` and `STF::Witness` associated type  now carry a `'static` bound. This constraint implies that only owned types can implement these traits, which I consider a reasonable requirement, though we're open to feedback on this matter.

### Other Noteworthy Changes

* Benchmarks now solely measure apply_slot. While this reduces granularity, it better represents our new design choice.
* Tests that previously checked whether a result is saved if end_slot isn't called are now ignored. I plan to transition these to a new testing strategy, where execution fails mid-way through applying blobs or even mid-transaction.

## Linked Issues
- Related to #408 

## Testing
Automated tests have been updated

## Docs
Doc strings have been updated

### CODE REVIEW

File: examples/demo-prover/host/src/main.rs
Comment (citizen-stig) [inline examples/demo-prover/host/src/main.rs]: Temporarely for faster debugging

File: examples/demo-prover/host/src/main.rs
Comment (preston-evans98) [inline examples/demo-prover/host/src/main.rs]: This seems risky. Let's compute it and store it in state instead. (I think Theo already has code to do this)

File: examples/demo-prover/methods/guest/src/bin/rollup.rs
Comment (preston-evans98) [inline examples/demo-prover/methods/guest/src/bin/rollup.rs]: nit: "has been"

File: module-system/sov-state/src/witness.rs
Comment (preston-evans98) [inline module-system/sov-state/src/witness.rs]: Question: Where did the new `'static` requirement come from?

File: rollup-interface/src/state_machine/stf.rs
Comment (preston-evans98) [inline rollup-interface/src/state_machine/stf.rs]: This represents a change in terminology. Before we called a da layer block a slot. Let's either change this to slot, or make sure we've updated all of our docs to use the new term.

File: rollup-interface/src/state_machine/stf.rs
Comment (preston-evans98) [inline rollup-interface/src/state_machine/stf.rs]: nit: outdated doc

File: rollup-interface/src/state_machine/stf.rs
Comment (preston-evans98) [inline rollup-interface/src/state_machine/stf.rs]: nit: outdated doc

File: rollup-interface/src/state_machine/stf.rs
Comment (preston-evans98) [inline rollup-interface/src/state_machine/stf.rs]: Let's remove this argument, since if the prover lies we don't have a way to detect it. We can still get access to the height inside the STF using the chain state module

**Review by preston-evans98 — COMMENTED** [review]
Looks pretty good overall. Just two issues  - first the height isn't authenticated, so the prover could like about it and cause a fork. And second the docs/specs still need to be updated with the new STF API.

File: module-system/sov-modules-stf-template/src/lib.rs
Comment (bkolad) [inline module-system/sov-modules-stf-template/src/lib.rs]: You can use `?` instead of pattern matching.

**Review by bkolad — APPROVED** [review]
LGTM once @preston-evans98 comments are addressed.

File: module-system/sov-state/src/witness.rs
Comment (citizen-stig) [reply module-system/sov-state/src/witness.rs]: Resolved!

File: module-system/sov-modules-stf-template/src/lib.rs
Comment (citizen-stig) [reply module-system/sov-modules-stf-template/src/lib.rs]: That's a good idea.

It didn't work originally, because of :

```
124 |               let batch_receipt = self.apply_blob(blob)?;
    |                                                        ^ cannot use the `?` operator in a method that returns `BlockResult<RootHash, SequencerOutcome, TxEffect, <<C as Spec>::Storage as Storage>::Witness>`
    |
    = help: the trait `FromResidual<Result<Infallible, ApplyBatchError>>` is not implemented for `BlockResult<RootHash, SequencerOutcome, TxEffect, <<C as Spec>::Storage as Storage>::Witness>`
    
```

But I can implement needed trait.

File: rollup-interface/src/state_machine/stf.rs
Comment (citizen-stig) [reply rollup-interface/src/state_machine/stf.rs]: Yes, and I think it is not blocking this PR, it will only be needed for soft confirmations PR.

File: examples/demo-prover/host/src/main.rs
Comment (citizen-stig) [reply examples/demo-prover/host/src/main.rs]: Yes, I've removed it from this PR, because it is not necessary now.

It will be added when blobs deferring will start to be used

File: module-system/sov-modules-stf-template/src/lib.rs
Comment (citizen-stig) [reply module-system/sov-modules-stf-template/src/lib.rs]: That's actually interesting observation.

If single blob in slot has failed to be applied, should we abort whole slot? If so, then `?` can be used.

But if slot should be continued, then just simple `.into()` is fine.

Comment (citizen-stig) [reply module-system/sov-modules-stf-template/src/lib.rs]: Actually in current logic, we keep going. @preston-evans98 please confirm if this is desired behaviour.

Also, I've get rid of `match` for something more compact.

File: rollup-interface/src/state_machine/mod.rs
Comment (preston-evans98) [inline rollup-interface/src/state_machine/mod.rs:30]: Question: do we still need this `'static`?

**Review by preston-evans98 — COMMENTED** [review]
Looks great!

**Review by preston-evans98 — APPROVED** [review]

File: rollup-interface/src/state_machine/mod.rs
Comment (citizen-stig) [reply rollup-interface/src/state_machine/mod.rs:30]: Yes, because BlobReaderTrait require it:

```rust
pub trait BlobReaderTrait: Serialize + DeserializeOwned + Send + Sync + 'static
```

and we need it there, as mentioned in PR description

File: module-system/sov-modules-stf-template/src/lib.rs
Comment (preston-evans98) [reply module-system/sov-modules-stf-template/src/lib.rs]: Yes, we want to keep going. Otherwise, it would be pretty easy to DOS the rollup by posting invalid blobs in every slot

**Pull request merged**
